### ISSUE TITLE
Zero-sized read in calculate_target_mtime not handled

### ISSUE THREAD

**Issue by gkoz** [rust-lang/cargo#1504]
This line can read zero bytes breaking the next line's assumption that `len() > 0`
https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_rustc/fingerprint.rs#L321

**Comment by alexcrichton** [comment]
Did this trigger an assertion for you? If so, is it reproducible?

**Comment by gkoz** [comment]
It triggered an arithmetic overflow in a debug build and an out of bounds assertion otherwise. The file it was trying to read was empty. After I made my cargo build ignore this problem (by pushing 0 into the vec) this doesn't seem to happen with stock cargo anymore.

**Comment by gkoz** [comment]
It's possible be that I had interrupted cargo previously leaving an empty `target/release/.fingerprint/$pkg-$hash/dep-bin-$pkg` behind which lead to
```
thread '<main>' panicked at 'assertion failed: index.end <= self.len()', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/slice.rs:518
```

**Comment by gkoz** [comment]
In case I wasn't clear enough, the condition in this issue can be reproduced by doing
```
> truncate -s0 target/debug/.fingerprint/*/dep*
> cargo build
thread '<main>' panicked at 'assertion failed: index.end <= self.len()', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/slice.rs:518
```
I have no idea what caused the file to become empty in my case but panicing seems like a poor way of handling this.

**Comment by alexcrichton** [comment]
I think this code has changed to the point that it's been fixed, but feel free to reopen if it still happens!

**Comment by gkoz** [comment]
This issue is still reproducible although I haven't since encountered whatever caused the fingerprint files to become empty. Patched in #2136.
